### ISSUE TITLE
Add TF-IDF artifact storage and vector-input similarity for predict() inference

### ISSUE THREAD

**Issue by woodwardmw** [sil-ai/aqua-api#560]
## Overview

Add storage and serving of TF-IDF encoder artifacts so that a fast `predict()` function can be added to the `tfidf` assessment in [`aqua-assessments`](https://github.com/sil-ai/aqua-assessments). This mirrors the train/predict paradigm already in place for `word-alignment` (eflomal) and `agent`.

Today:

- `assess()` fits two `TfidfVectorizer`s (word + char_wb) and a `TruncatedSVD`, and computes a 300-dim PCA-reduced vector for every verse in the corpus.
- The runner pushes those per-vref vectors to `POST /v3/assessment/{id}/tfidf-vectors`.
- `GET /v3/tfidf_result?assessment_id=&vref=` returns nearest-neighbour vrefs to a vref **already in the corpus**.

What's missing for `predict(text)`:

1. The fitted **encoder artifacts** (word vectorizer, char vectorizer, SVD) are discarded after `assess()` — there is no way to encode a new piece of text into the same 300-dim space.
2. The similarity endpoint is keyed by `vref`, not by an arbitrary vector — so even after encoding a new text, there is no way to ask "which corpus verses are most similar to this vector?".

This issue covers the API-side changes. The corresponding `predict()` implementation in `aqua-assessments` is a follow-up.

---

## Scope

**In scope (this issue, aqua-api):**

- Alembic migration: 3 new tables for TF-IDF artifacts.
- SQLAlchemy models.
- 3 new REST endpoints: push artifacts, fetch artifacts, similarity-by-vector.
- Tests: round-trip, by-vector similarity, backward compatibility of existing `GET /v3/tfidf_result`.

**Out of scope (follow-up issues):**

- `predict()` function in `aqua-assessments/assessments/tfidf/app.py`.
- Modifying `assess()` in `aqua-assessments` to actually push artifacts (separate PR).
- Bulk delete / re-run / admin endpoints.
- Cross-assessment vector lookup (vector against the union of multiple corpora).

---

## 1. Why JSON/binary, not pickle

Pickled `sklearn` objects break across version upgrades (scikit-learn frequently changes private attributes of `TfidfVectorizer` and `TruncatedSVD`). Storing the inner state explicitly is more robust:

- `TfidfVectorizer` → `vocabulary_` (token → index dict), `idf_` (float array of length `vocab_size`), plus the constructor params (`analyzer`, `ngram_range`, `lowercase`, `max_df`, `min_df`).
- `TruncatedSVD` → `components_` (`n_components × n_features_combined` float matrix), `n_components`.

At inference time, `aqua-assessments` rehydrates these into fresh sklearn objects (the inverse operation is straightforward: assign to `vocabulary_` / `idf_` / `components_` on freshly-constructed instances).

---

## 2. DB schema

Three new tables in `aqua-api/database/models.py`. All link to `assessments`.

### 2.1 `tfidf_artifact_runs`

One row per assessment that produced TF-IDF artifacts. The "header" row.

```python
class TfidfArtifactRun(Base):
    __tablename__ = "tfidf_artifact_runs"

    assessment_id = Column(
        Integer,
        ForeignKey("assessments.id", ondelete="CASCADE"),
        primary_key=True,
    )
    source_language = Column(String(8), nullable=True)  # ISO 639-3 if known
    n_components = Column(Integer, nullable=False)       # 300 today
    n_word_features = Column(Integer, nullable=False)    # |word vocab|
    n_char_features = Column(Integer, nullable=False)    # |char vocab|
    n_corpus_vrefs = Column(Integer, nullable=False)
    sklearn_version = Column(Text, nullable=False)       # e.g. "1.6.1" — for diagnostics
    created_at = Column(TIMESTAMP, default=func.now())

    __table_args__ = (
        Index("ix_tfidf_artifact_runs_lang", "source_language"),
    )
```

### 2.2 `tfidf_vectorizers`

Two rows per assessment (one for `word`, one for `char_wb`).

```python
class TfidfVectorizer(Base):
    __tablename__ = "tfidf_vectorizers"

    id = Column(Integer, primary_key=True)
    assessment_id = Column(
        Integer,
        ForeignKey("tfidf_artifact_runs.assessment_id", ondelete="CASCADE"),
        nullable=False,
    )
    kind = Column(Text, nullable=False)              # "word" | "char"
    vocabulary = Column(JSONB, nullable=False)       # {token: index}
    idf = Column(JSONB, nullable=False)              # [float, ...] length = vocab_size
    params = Column(JSONB, nullable=False)           # {analyzer, ngram_range, lowercase, max_df, min_df}

    __table_args__ = (
        UniqueConstraint("assessment_id", "kind", name="uq_tfidf_vectorizer_assessment_kind"),
    )
```

### 2.3 `tfidf_svd_components`

The `components_` matrix is the heaviest artifact — for a 300-dim model with ~60k combined features that's ~70MB serialized as JSON. Two storage options, picking one:

**Option A (recommended): bytea + numpy `.npy`**

```python
class TfidfSvd(Base):
    __tablename__ = "tfidf_svd"

    assessment_id = Column(
        Integer,
        ForeignKey("tfidf_artifact_runs.assessment_id", ondelete="CASCADE"),
        primary_key=True,
    )
    n_components = Column(Integer, nullable=False)
    n_features = Column(Integer, nullable=False)
    components_npy = Column(LargeBinary, nullable=False)   # np.save() bytes of components_
    dtype = Column(Text, nullable=False, default="float32")  # store as float32 to halve size
```

**Option B: row-per-component**

`tfidf_svd_components(assessment_id, component_idx, weights JSONB)` — 300 rows of ~60k floats each. Cleaner schema but ~50% larger on disk and slower to round-trip.

I recommend **A** for size/speed; **B** if the team prefers no `bytea` columns. Either way the API surface is identical.

---

## 3. Alembic migration

Single migration: `add_tfidf_artifact_tables.py`. Independent of existing tables; no data migration needed.

---

## 4. API endpoints

New route file: `aqua-api/routes/v3/tfidf_artifact_routes.py`. Same auth pattern as existing `/v3/assessment/eflomal/*` routes.

### 4.1 `POST /v3/assessment/{assessment_id}/tfidf-artifacts`

Single fat write — everything in one transaction, mirroring the morpheme tokenizer commit endpoint.

**Request body:**

```json
{
  "source_language": "swh",
  "n_components": 300,
  "n_corpus_vrefs": 31102,
  "sklearn_version": "1.6.1",
  "word_vectorizer": {
    "vocabulary": {"and": 0, "the": 1, "...": 2},
    "idf": [1.23, 0.87, 4.51],
    "params": {
      "analyzer": "word",
      "ngram_range": [1, 2],
      "lowercase": true,
      "max_df": 0.12,
      "min_df": 2
    }
  },
  "char_vectorizer": {
    "vocabulary": {" th": 0, "the": 1, "he ": 2},
    "idf": [3.11, 0.42, 2.07],
    "params": {
      "analyzer": "char_wb",
      "ngram_range": [3, 6],
      "lowercase": true,
      "max_df": 0.3,
      "min_df": 2
    }
  },
  "svd": {
    "n_components": 300,
    "n_features": 60000,
    "dtype": "float32",
    "components_b64": "<base64-encoded np.save() bytes>"
  }
}
```

**Server logic:**

1. Validate `assessment_id` exists in `assessments` and its `type == "tfidf"`. 422 if not.
2. Upsert `tfidf_artifact_runs` row (delete-and-replace on conflict).
3. Upsert two `tfidf_vectorizers` rows (`word`, `char`).
4. Upsert one `tfidf_svd` row (decode `components_b64` → store as `LargeBinary`).
5. All in one transaction. Return 200 with:

```json
{
  "assessment_id": 1234,
  "n_word_features": 12000,
  "n_char_features": 48000,
  "components_bytes": 72000000
}
```

**Notes:**

- The body can be large (~70MB+). Set the API's max request body accordingly. Compression (gzip) on the client side is encouraged — FastAPI handles `Content-Encoding: gzip` transparently if the middleware is enabled (currently not — add it).
- Idempotency: re-posting replaces all artifacts for that assessment.

### 4.2 `GET /v3/assessment/tfidf/artifacts`

Mirror of `/v3/assessment/eflomal/results`. Fetch by `assessment_id` or by `source_language` (latest run).

**Query params:**

- `assessment_id` (int, optional)
- `source_language` (str, optional) — used only if `assessment_id` is omitted; returns the most recent run for that language

Exactly one of the two must be provided. 422 otherwise.

**Response:** identical shape to the POST body in §4.1, plus `created_at`.

**404** if no matching artifact run exists.

### 4.3 `POST /v3/tfidf_result/by_vector`

Vector-input similarity search. The companion to the existing vref-keyed `GET /v3/tfidf_result`.

**Request body:**

```json
{
  "assessment_id": 1234,
  "vector": [0.012, -0.045, ...],   // length must equal n_components stored for this assessment
  "limit": 10,
  "reference_id": null
}
```

**Server logic:**

1. Validate `assessment_id`, fetch its `n_components`. 422 if vector length mismatch.
2. Compute cosine similarity (or dot product if vectors are already L2-normalized — they are, post-SVD; document this) against all rows in `tfidf_vectors` for this assessment.
3. Return top-`limit` as `List[TfidfResult]` — identical schema to the existing `GET /v3/tfidf_result` response.

**Performance note:** for 31k vrefs × 300 dims, a brute-force matmul is ~10ms in numpy and totally fine. If we ever scale to multiple revisions per query, consider `pgvector`. Out of scope here.

The existing `GET /v3/tfidf_result?vref=` endpoint stays unchanged; internally it can be refactored to call the same scoring core as `by_vector` after looking up the source vector by vref.

---

## 5. Tests

Add to `aqua-api/test/`. At minimum:

### 5.1 Round-trip

```python
def test_tfidf_artifact_round_trip(client, tfidf_assessment):
    body = make_fake_artifacts(n_components=300, n_word_features=10, n_char_features=20)
    resp = client.post(f"/v3/assessment/{tfidf_assessment.id}/tfidf-artifacts", json=body)
    assert resp.status_code == 200
    assert resp.json()["n_word_features"] == 10

    resp = client.get(f"/v3/assessment/tfidf/artifacts?assessment_id={tfidf_assessment.id}")
    out = resp.json()
    assert out["word_vectorizer"]["vocabulary"] == body["word_vectorizer"]["vocabulary"]
    assert out["svd"]["n_features"] == 30
    # numeric equality on the components matrix
    assert decode_components(out["svd"]) == pytest.approx(decode_components(body["svd"]))
```

### 5.2 By-vector similarity

Seed `tfidf_vectors` with three known unit vectors at vrefs `GEN 1:1`, `GEN 1:2`, `GEN 1:3`. POST a vector close to `GEN 1:2`. Assert top-1 is `GEN 1:2`.

### 5.3 Vector length validation

POST `by_vector` with `len(vector) != n_components`. Assert 422.

### 5.4 Backward compatibility

Existing `GET /v3/tfidf_result?vref=` keeps working unchanged after the refactor.

### 5.5 Idempotency

POST artifacts twice. Assert second call replaces, not duplicates (still exactly one row in each of the three tables).

### 5.6 Latest-by-language

Two artifact runs for the same `source_language`. `GET /v3/assessment/tfidf/artifacts?source_language=swh` returns the more recent one.

---

## 6. File layout

```
aqua-api/
├── database/
│   └── models.py                    # + TfidfArtifactRun, TfidfVectorizer, TfidfSvd
├── alembic/versions/
│   └── <timestamp>_add_tfidf_artifact_tables.py
├── routes/v3/
│   └── tfidf_artifact_routes.py     # 3 new endpoints
├── app.py                           # mount the new router; enable gzip middleware
└── test/
    └── test_tfidf_artifact_routes.py
```

---

## 7. Follow-up work (separate issues / PRs)

**aqua-assessments:**

1. `assessments/tfidf/app.py` — modify `assess()` to extract vectorizer state and SVD components and POST them to `/v3/assessment/{id}/tfidf-artifacts` after vector upload.
2. `assessments/tfidf/app.py` — add `predict(text: str, assessment_id: int) -> {vector, neighbours}`:
   - Fetch artifacts via `GET /v3/assessment/tfidf/artifacts`.
   - Rehydrate `TfidfVectorizer`s + `TruncatedSVD` from the JSON / npy payload.
   - Encode `text` → 300-dim vector.
   - POST to `/v3/tfidf_result/by_vector` for nearest-neighbour vrefs.
3. `assessments/tfidf/app_test.py` — round-trip test through Modal.

**aqua-api (later, if needed):**

- `pgvector` integration for similarity if brute-force scoring becomes a bottleneck.
- Optional `POST /v3/tfidf_result/by_text` — server-side encoding, eliminating the round-trip. Only worth it if multiple clients want this.

---

## 8. Reference

- aqua-assessments tfidf app: `assessments/tfidf/app.py`
- Runner code that pushes per-vref vectors today: `runner/runner/app.py:344`
- Analogous artifact pattern (eflomal): `assessments/word_alignment/eflomal_steps/push_results.py`, `assessments/word_alignment/eflomal_steps/artifacts.py`, and the `POST /v3/assessment/{id}/eflomal-{dictionary,cooccurrences,target-word-counts}` endpoints
- Analogous design issue: sil-ai/aqua-api#520 (morpheme tokenizer storage)
